internal/testutils/networking/l2_segment_linux.go

```diff
@@ -0,0 +1,339 @@
+package networking
+
+import (
+	"fmt"
+	"net"
+	"net/netip"
+	"os"
+	"runtime"
+	"syscall"
+	"testing"
+
+	"github.com/docker/docker/libnetwork/testutils"
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+	"gotest.tools/v3/assert"
+)
+
+// Netns represents a network namespace that might be named, and thus mounted under /run/netns.
+type Netns struct {
+	handle netns.NsHandle
+	name   string
+}
+
+// CurrentNetns returns an unnamed Netns for the current network namespace this thread is living in. If it fails to
+// get a handle for the current net namespace, it bails out.
+func CurrentNetns(t *testing.T) Netns {
+	t.Helper()
+
+	curNs, err := netns.Get()
+	assert.NilError(t, err)
+
+	return Netns{handle: curNs}
+}
+
+// NewNamedNetns creates a new named Netns and then switch back to the original net namespace. It bails out if it fails.
+func NewNamedNetns(t *testing.T, name string, reusePrevious bool) Netns {
+	t.Helper()
+
+	var err error
+	newNs := Netns{name: name}
+	newNs.handle, err = netns.GetFromName(newNs.name)
+	if err != nil && !os.IsNotExist(err) {
+		assert.NilError(t, err)
+	}
+	if err == nil {
+		if reusePrevious {
+			return newNs
+		}
+		t.Fatalf("Network namespace %s already exists whereas TEST_REUSE_L2SEGMENT is empty.", newNs.name)
+	}
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	origNs, err := netns.Get()
+	assert.NilError(t, err)
+
+	newNs.handle, err = netns.NewNamed(newNs.name)
+	assert.NilError(t, err)
+
+	err = netns.Set(origNs)
+	assert.NilError(t, err)
+
+	return newNs
+}
+
+// Set joins the current goroutine to this Netns. The caller has to make sure the OS thread is locked.
+func (ns Netns) Set() error {
+	if !ns.handle.IsOpen() {
+		return nil
+	}
+	if err := netns.Set(ns.handle); err != nil {
+		return fmt.Errorf("could not switch to netns %s: %w", ns.String(), err)
+	}
+	return nil
+}
+
+// InNetns executes the provided fn in the Netns and then switches back to the original netns. If it fails to restore
+// the original netns, it marks the test as failed and stops its execution. For that reason, InNetns() should be run in
+// the test goroutine.
+func (ns Netns) InNetns(t *testing.T, fn func() error) error {
+	t.Helper()
+
+	orig := CurrentNetns(t)
+	runtime.LockOSThread()
+	defer func() {
+		if err := netns.Set(orig.handle); err != nil {
+			t.Fatalf("could not switch back to the original netns: %v", err)
+		}
+		runtime.UnlockOSThread()
```

Comment:
It's ok to skip the `runtime.UnlockOSThread()` if we fail? (I guess it is if we `os.Exit()`